BUG: Fixes for building on Cygwin.

DWesl opened the pull request:
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

Changes needed to get numpy to compile on Cygwin, then a few changes to try to reduce the number of test failures due to extension module overlap, then marking differences I think are due to floating-point implementation or fork() failures.

Someone should probably check whether the list of ignored floating-point failures is reasonable.
Most of those have been around since 1.16 a while back.

DWesl committed:
BLD: Move the configuration defines in Python.h earlier.

Leaving it later can lead to some headers getting confused, because
they're included with one set of feature flags, but the headers they
depend on were included with a different set of feature flags.

DWesl committed:
DEV: Try to rebase f2py-created modules for fewer fork() failures in rebase.

Cygwin needs each DLL to have a unique address for fork() emulation to
work.  I'm hoping that calling rebase on a complete list of modules
compiled in this session, plus those installed globally, will allow
the test suite to get an accurate result for more tests.

DWesl committed:
TST: Mark the tests that seem to fail due to different floating-point on cygwin.

Most likely I should be testing for newlib (the C runtime, roughly
takes the place of glibc, I think), not cygwin (the emulation layer on
Windows), but I have no idea how to do that from within python.  If
someone working on embedded systems runs into this issue, this
hopefully gives them some idea where to start.

DWesl committed:
TST: Try to mark the tests still suseptible to fork() failures on cygwin.

See two commits back for an explanation of why fork() fails on cygwin.
Alternately, see the much better explanation at:
https://cygwin.com/cygwin-ug-net/highlights.html#ov-hi-process
with hints on error messages and workarounds at:
https://cygwin.com/faq.html#faq.using.fixing-fork-failures

DWesl commented:
The failure in Travis seems to be because it couldn't install gfortran, which seems unrelated.

seberg commented:
The header change seems trivial, the distutils change I do not know, maybe @rgommers can have a look at some point.
Some of the test changes look like serious precision issues, I am wondering if those are cases where we have an alternative implementation, but do not blacklist the implementation shipped by cygwin.  The blacklists are in `numpy/core/src/common/npy_config.h` (not sure if/how it would apply).

@embray you seems to have looked at cygwin related things before. If you have time, any input is appreciated!

mattip commented:
Looking at `npy_config.h` is a good idea since it may be setting things wrong by blacklisting things that may be OK on a newer mingw. It would be nice if we could record the choices made there in `__config__` somehow.

DWesl commented:
To clarify, this is Cygwin, not MinGW.  I can try to cross-compile or check wheels from somewhere if necessary, but nothing here is related to MinGW.

As an interesting side note, compiling latest master works just fine, and seems to have fewer failures than this branch.  I'm suspecting I did something wrong with the `rebase` patch.  

I'll try to check the problematic floating-point functions manually and add problematic ones to `npy_config.h`

embray commented on numpy/distutils/fcompiler/__init__.py:
```diff
@@ -681,6 +684,14 @@ def link(self, target_desc, objects,
+                rebase_args = ['/usr/bin/rebase', '--database',
+                               '--oblivious']
+                rebase_args.extend(COMPILED_MODULES)
+                self.spawn(rebase_args)
```
I'm going to see if I can find it, but at some point I wrote (for Sage) some code similar to this but a little more advanced which tried to find an open memory range to rebase compiled modules to.  I don't recall why exactly but just using `rebase` directly wasn't always working for me (sometimes the rebase database wasn't always available).  I don't recall the details though.

Of course, in general this is the right idea.

DWesl replied:
The tests run better without this than with it at the moment.  I think the problem is not including the numpy dlls themselves in the `rebase`.  There is a way to list those using pip, which would allow be to include those in the list passed to `rebase`, but it doesn't look like that's needed on 64-bit Cygwin at the moment.  I'll remove this code soon.

DWesl replied:
I removed this code.

embray commented:
I'm happy to take a look at this, but could you provide a brief overview of what this fixes?  I haven't personally had any problems lately with *compiling* Numpy on Cygwin.  Is this fixing any build issues?  Otherwise it seems to be mostly concerned with fixing the test suite, and setting a number of tests to xfail.  I admit I haven't tried running the full test suite on Cygwin in a while.

embray commented on numpy/core/tests/test_numeric.py:
```diff
@@ -585,6 +585,9 @@ def test_errobj_noerrmask(self):
 class TestFloatExceptions:
+    @pytest.mark.xfail(sys.platform == 'cygwin',
+                       reason='FPE support incomplete',
+                       strict=True)
```
Could you elucidate what you mean by "FPE support incomplete"?  What FPE APIs are unavailable on Cygwin?  I seem to recall more recent versions of Cygwin have pretty good support for floating point exceptions.

DWesl replied:
I know this test failed a few months ago; that may have been one of the things that got fixed in the last six months or so.  Unfortunately "this test involves FPEs and fails" is about the sum of my knowledge on the subject.

DWesl replied:
This may only apply to the test documented a few functions down; since this is applied to an assert function rather than a test function, it may not do anything useful.  It may be a good idea to drop this mark and see what happens.

DWesl replied:
I dropped this mark and have noticed no new failures.

eric-wieser commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -667,9 +667,13 @@ def _test_abs_func(self, absfunc):
             x = tp(np.finfo(tp).min)
             assert_equal(absfunc(x), -x.real)
 
+    @pytest.mark.xfail(sys.platform == 'cygwin',
+                       reason='abs(np.complex256.max) overflows')
     def test_builtin_abs(self):
         self._test_abs_func(abs)
 
+    @pytest.mark.xfail(sys.platform == 'cygwin',
+                       reason='abs(np.complex256.max) overflows')
     def test_numpy_abs(self):
         self._test_abs_func(np.abs)
```
Rather than doing this, it would be good to make `_test_abs_func` use `pytest.mark.parametrize` so that we can only xfail `complex256`

DWesl replied:
Done.  It would be nice to have something similar for the branch cut tests, but there's too much variation to make that simple.

DWesl replied:
Now done in a way that doesn't crash if the platform doesn't have a 256-bit complex type.

DWesl commented:
A few months ago (https://github.com/numpy/numpy/issues/14787#issuecomment-586731280, though I think I had noted the problem earlier), numpy wasn't compiling on Cygwin unless I made `Python.h` the first file to be `#include`d in several files.  That problem appears to be gone now.  That is the first commit.

I was getting a lot of `fork()` failures in tests at the time, and thought running `rebase` as part of `f2py` would help.  I think it might have at first, but I can now run the tests with only a few `fork()` failures, without any changes to NumPy master, so that problem also appears to be gone now.  That is the core of the second commit.  I also changed several program names, which make sense when compiling only for Cygwin, but not for NumPy master.

Several tests failed and still fail due to what seemed like tiny floating-point differences, sometimes nonexistent differences as the branch cut tests don't always report which side of the branch they're trying to test.  My first approach was to ignore them, as I saw no relevant differences.  That is the third commit.  Earlier comments in this thread directed me to `npy_config.h` to mark the failing functions, so that NumPy could use its own implementations.  I have started to work through that, but get some test failures on functions I have marked, so I may not be able to get all of the functions that should be there.  I will post that work soon.

There were still tests with `fork()` failures.  I marked them as `xfail`, because I cared more about whether NumPy worked in general rather than whether my environment was amenable to subprocesses.

A question for the maintainers: should I try to include the modified `rebase` code here?  If NumPy is installed through the Cygwin package manager, `rebase` is called automatically.  The `rebase` code would be primarily useful for people who create many `f2py` modules or install NumPy on their own.
Another possibility for the latter case is to add a Cygwin section to the Install docs telling people how to do this on their own and describing when it's necessary.

DWesl commented:
@embray 
Cygwin 3.1.4, GCC 9.3.0

Out of curiosity, what AVX-related errors are you seeing?  I ran into some a few months ago and reported those in #14787, where other people mentioned possible solutions.

DWesl committed:
DEV: Revert executable renames for F77/G77 compiler.

These changes made some sense when compiling only for cygwin.  They do
not belong in NumPy master.

DWesl committed:
BLD: Add functions that fail tests on Cygwin to npy_config.h.

I don't know if the list should be the same as one of the Windows
lists, or if this should instead be a Newlib list.  I know how to make
a Cygwin list, so I did that.  It didn't always work, though.

DWesl committed:
TST: Describe what I know about the FPE test failure.

The generic "this doesn't work" was confusing people.  Specifying what
"this" is and what happed works better.

DWesl committed:
STY: A few largely cosmetic changes.

Mark tests suddenly passing, and also document how I expect them to
fail.

DWesl committed:
DEV: Stop trying to rebase f2py modules.

This was actually making the situation worse, I think because I forgot
to include the NumPy dlls themselves.  Removing this made for many
fewer fork() failures when I tried it.

`python3 -m pip show numpy --files | grep dll`
will give a list of dlls, with at least a relative path.  This can be
done in python with subprocess (will require adding pip as a runtime
dependency), but I'm not sure that's in scope here.

DWesl committed:
Remove xfails from tests that now pass.

DWesl committed:
Use parametrize for the dtypes in the abs tests.

DWesl committed:
Remove xfail mark from an assert function.

These should primarily be on test functions.

DWesl committed:
Remove references to np.complex256

Unconditional access to np.complex256 triggers an error in test collection on platforms where the dtype does not exist.
Find another way to test the dtype for equality.

DWesl committed:
Remove reference to np.complex256 in the other test

I changed this for the one test, but forgot to do so for the other.  Hopefully this actually fixes the problem.

DWesl commented:
As near as I can tell, the Travis failure is due to being unable to download python3.7 on s390x.  I don't think that's related to my changes here.

DWesl commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -84,7 +84,7 @@ def version_match(self, version_string):
         return v[1]
 
-    possible_executables = ['g77', 'f77']
+    possible_executables = ['g77', 'f77', "gfortran"]
```
```suggestion
    possible_executables = ['g77', 'f77']
```

DWesl committed:
Don't add `gfortran` to the list of GNU77 compilers

Presumably whoever wrote this code had reasons for leaving this out, and making a separate class for `gfortran`

DWesl commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -287,7 +289,7 @@ def version_match(self, version_string):
         return v
 
-    possible_executables = ['gfortran', 'f95']
+    possible_executables = ['gfortran', 'f95', '/usr/bin/gfortran']
```
```suggestion
    possible_executables = ['gfortran', 'f95']
```
This was originally to help f2py find my Fortran compiler, but that doesn't seem to be an issue anymore.  Should I drop this change?

DWesl committed:
Include only 'gfortran' in gnu fcompiler, not absolute version

Including both 'gfortran' and '/usr/bin/gfortran' should be redundant, and I'm not sure even my platform ever needed this.  Remove it.

DWesl reviewed (commented):
I think this was what I found when trying to re-derive this changeset.

I think I dropped all the extraneous parts of the patch.  What else do I need to do to get this in?

mattip commented on numpy/random/src/distributions/random_hypergeometric.c:
```diff
@@ -1,6 +1,7 @@
+#include "Python.h"
```
Nothing in this file should require `Python.h`

DWesl replied:
```suggestion
```
NumPy compiles without this change on latest Cygwin.

mattip commented on numpy/core/src/common/npy_config.h:
```diff
@@ -81,6 +81,55 @@
+#undef HAVE_CATANHL
+
+#endif
+
```
Are these due to cygwin using an old glibc? If so, maybe we can tie it to the glibc version like below.

DWesl replied:
Cygwin is based on Newlib, not GNU C-library.  If glibc is something different I can look into checking that.

The list of problem functions here looks very similar to the one for MSVC and MinGW; I think acos and friends are the only ones in the Windows list but not the Cygwin one.  I may consolidate those lists.

DWesl replied:
The `__CYGWIN__` macro unfortunately does not include a version.

The `modfl` code in cygwin is taken from the MinGW-w64 project, so the inverse trigonometric functions working in Cygwin may mean that they're working on Windows and can be removed from that list.  I don't have a Windows build environment, so it will be a while before I could try that.

mattip reviewed (commented):
I am a little concerned about the quality of the resulting NumPy. Are we sure at this point, given the skipped tests, that we can sign off on cygwin as a "supported" compiler?

DWesl reviewed (commented):
@mattip Which skips are you worried about?  The forking issues are a reasonably well-known problem within the Cygwin community, and the solutions similarly so.  I am less qualified to talk about the difficulties with branch cuts or overflow/nan behavior, so I will leave decisions on those skips to the maintainers.

embray commented:
As far as I can tell the answer is "no", but do you plan for this to include a fix for #14787?  I also found that adding

```
CFLAGS="-fno-asynchronous-unwind-tables
-ffixed-xmm16 -ffixed-xmm17 -ffixed-xmm18 -ffixed-xmm19
-ffixed-xmm20 -ffixed-xmm21 -ffixed-xmm22 -ffixed-xmm23
-ffixed-xmm24 -ffixed-xmm25 -ffixed-xmm26 -ffixed-xmm27
-ffixed-xmm28 -ffixed-xmm29 -ffixed-xmm30 -ffixed-xmm31"
```

worked around the issue for me.  So perhaps this could be added automatically by the build, at least for those GCC versions where the issue applies (GCC < 8.4).

DWesl commented:
@embray It would be a good idea, but I'm not sure where to put that test.

The only thing coming to mind is to test for Cygwin in `setup.py` and modify `os.environ["CFLAGS"]`, but I know NumPy already has code for flags somewhere, so this should probably go there.  It would be best if I could also test the compiler version as well, but that's not as big a concern.

embray commented:
Yeah, it's not necessary to use `os.environ`.  There are places in the `setup.py`'s to provide additional compiler flags for building individual modules.  These flags are I think necessary just for `loops.c` and can be added on a conditional check.  I'm not 100% sure if there's already code in there to check compiler version but I think there might be.

mattip commented:
> I'm not 100% sure if there's already code in there to check compiler version but I think there might be.

xref pr #16619 which adds `-std=c99` dependent on compiler version

mattip commented:
> The list of problem functions here looks very similar to the one for MSVC and MinGW; I think acos and friends are the only ones in the Windows list but not the Cygwin one. I may consolidate those lists.

thanks

mattip commented:
> The forking issues are a reasonably well-known problem within the Cygwin community

So how do people with a cygwin build avoid this problem in things like multiprocess ?

DWesl commented:
> > The forking issues are a reasonably well-known problem within the Cygwin community
> 
> So how do people with a cygwin build avoid this problem in things like multiprocess ?

From most to least important or impactful, in my highly subjective opinion:
1. Ensure dlls are set up to avoid address collisions (default post-install will do this for `/usr/lib/python${ver}/site-packages`; you can change this to also do `~/.local/lib/python${ver}/site-packages` or a virtual environment at all future package updates)
2. Avoid dll injection from anti-malware software (known problems are called [BLODA](https://cygwin.com/acronyms/#BLODA); solutions include whitelisting the Cygwin installation directory)
3. Use `threading` instead of `multiprocessing` or `subprocess`
4. Set process creation to use `spawn` instead of `fork`/`exec`
5. Use the 64-bit release rather than the 32-bit release.  The 64-bit version has a larger address space and correspondingly lower risk of address collision.

The rebase failures are something else that seem to have vanished on my (64-bit) installation over the last few months.  I can restrict the `xfail`s to 32-bit Cygwin installations only (`sys.maxsize >> 30 == 1` should work), or I can try to replace every call to `subprocess`, `multiprocessing`, or `distutils.spawn` with appropriate 
```python
try:
    call_involving_fork()
except BlockingIOError:
    if sys.platform == "cygwin":
        pytest.xfail(reason="fork() fails randomly on cygwin.")
    raise
```
blocks to minimize the places where pytest looks for these.  This has the advantage of being able to catch failures in the setup methods, which are somewhat difficult to track down using the tracebacks in pytest.

DWesl committed:
BLD: Remind GCC that .seh_savexmm fails for xmm16-31.

Workaround for GCC bug 65782.
Fixes numpy issue 14787.

mattip commented:
Thanks for the extensive reply. 
> I can restrict the xfails to 32-bit Cygwin installations only

This sounds reasonable. You could add an `IS_32BIT_CYGWIN` to `testing/private/utils.py` next to `IS_PYPY`

DWesl commented:
> If that doesn't work and failures occur on 64 bit as well

At present, things only fail on 64bit if I try to prevent failures.  I've changed the xfail condition over and will push the updates once I convince the tests to run.

DWesl committed:
BLD: Fix code checking for old GCC on cygwin.

I forgot the function returned bytes, not str.
I also accidentally passed an extra argument.

DWesl committed:
TST: Only mark fork()-using tests xfail on 32-bit cygwin.

The problem seems to have gone away on 64-bit.

DWesl committed:
BLD: Fix flags added to compiler line for old GCC on MS.

It may be worth broadening the platform list, since I believe the
error this tries to fix is in the MS ABI/CRT rather than in the Cygwin
runtime.

DWesl committed:
Merge branch 'master' into cygwin-fixes

eric-wieser commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -646,32 +646,67 @@ def test_result(self):
+                        raises=RuntimeWarning,
+                        strict=True,
+                    )
+                    if dtype().dtype.str.endswith("c32") else ()
```
```suggestion
                    if dtype == np.complex256 else ()
```

DWesl replied:
`np.complex256` doesn't exist on all targets, causing test failures.
See b530f99 and c11ee6f for when this was changed to use "c32" rather than `np.complex256`

eric-wieser replied:
` == np.clongdouble` would also be an option.

Does cygwin always have complex256?

DWesl replied:
My (64-bit) Cygwin always has complex256.  I can hide that comparison behind an `sys.platform == "cygwin" and` so it is only reached on Cygwin.

eric-wieser replied:
Forgive me if I missed this conversation - this sounds like the problem is that `long double` on cygwin is actually just `double` - I've seen this behavior in WSL, and it came down to the fpu being in low precision mode, and downcasting `long double` to double before each operation. Does that sound plausible?

DWesl replied:
```python
>>> np.array([10 + 10j], dtype=np.complex256) ** 4000
array([1.1481307e+4602+1.23982657e+4586j], dtype=complex256)
>>> np.abs(_)
array([inf], dtype=float128)
```
I don't think the problem is `long double == double`.

That number should have an absolute value representable as a long double, but it would appear Cygwin has trouble finding that long double.  (`LDBL_MAX == 1.1e4932` on my system).

DWesl committed:
TST: Use usual names for the dtypes in xfail mark

Short-circuiting is lovely.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -8262,6 +8262,9 @@ class MyAlwaysEqualNew(MyAlwaysEqual):
 
+@pytest.mark.xfail(sys.platform == 'cygwin',
+                   reason='abs(inf+nanj) short-circuits to inf',
+                   strict=True)
```
Is this just the `clongdouble` loop iteration below failing?

eric-wieser replied:
It would be great to `parameterize` this test to make the xfail more precise.

DWesl replied:
```python
>>> np.abs(np.array([complex(np.inf, np.nan)], np.complex64))
array([inf], dtype=float32)
>>> np.abs(np.array([complex(np.inf, np.nan)], np.complex128))
array([inf])
>>> np.abs(np.array([complex(np.inf, np.nan)], np.complex256))
array([nan], dtype=float128)
```
As near as I can tell, `clongdouble` is the only one getting this right.

It looks relatively easy to parametrize.  I'll look into that.

DWesl committed:
TST: parametrize test_npymath_complex.

Allows more precise xfail.

DWesl commented:
For about the past three weeks, I keep getting segfaults when I try to run tox on my machine.

This is probably something with my computer, and is highly unlikely to be a NumPy issue, but it will slow my progress on testing until I figure out how to work around this.  
I think I've seen this when re-compiling the release I have installed, though the installed tested (and tests) just fine.
I just checked 1.17.5, which I'm pretty sure tested fine two months ago, and it now segfaults in the same place.

Has anyone else run into this issue, or at least know how to work around it?  
I included the more relevant lines of the error below.

```
Fatal Python error: Segmentation fault

Current thread 0x0000000800000010 (most recent call first):
  File "~/src/DWesl/numpy/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_ufunc.py", line 2013 in test_ufunc_types
... (Pytest internals)
```

charris commented:
Sometimes I see strange errors like that, make sure to run `git clean -xdfq` before compiling and testing, that usually fixes the problem. Why aren't you using `python runtests.py`?

DWesl commented:
I forget it's there and try to run pytest directly, which fails because I'm in the numpy repository root.
Tox lets me forget about that.  I'll try `runtests.py`.

DWesl commented:
I think the one Travis failure is due to 'mkl' appearing in a tempfile name.

The overzealous code changing the temporary directory is probably:
numpy/distutils/tests/test_system_info.py:284:
```python
            # Also, the values will be taken from a section named '[DEFAULT]'
            with open(cfg, 'r') as fid:
                dflt = fid.read().replace('mkl', 'DEFAULT')
```
Putting brackets around everything may prevent the issue from recurring, as might telling the replace method that it should only replace 'mkl' once.

I believe this is everything that should be marked `xfail`.  
`isinstance(np.array(), collections.Hashable)` succeeds, but that should probably be fixed rather than ignored.

DWesl committed:
TST: Try to keep testing code from changing tempfile name.

I think I saw a test failure because the generated tempfile name
contained the string "mkl".  I think what the code is trying to do is
to change the "[mkl]" section to a "[DEFAULT]" section without
changing any of the contents of that section.  I think this change
should do the trick.

DWesl commented:
Attempting to run the failing tests line-by-line in gdb indicates `np.modf(np.ones((3, 3), dtype="g"))` segfaults.
Unfortunately, this does not do so in a way that tells gdb.

New numpy obtained from `build/lib.cygwin-*-x86_64-3.6`.

DWesl committed:
BLD: Mark modfl as problematic on cygwin.

The implementation currently segfaults.
I'll report this to the developers soon;
hopefully it can be re-enabled soon.

DWesl committed:
TST: Clarify the xfail messages for branch cuts.

I think these are the tests where the "branch cut code doesn't use
sign of zero" became relevant.

DWesl committed:
BLD: Remove probably-unnecessary include.

This was trying to ensure the definition of symbols that depend on
feature macros set in "Python.h".  That issue seems to have resolved
itself.

DWesl commented:
There's a problem with Cygwin `modfl`, which has been reported to Cygwin.  In the meantime, `npy_config.h` has been changed to reflect that `modfl` segfaults.

I also tried to prevent `test_system_info.py` from changing the tempfile name in addition to the section name and incorporated the suggestions from review comments above.

embray commented:
Let me know if there are still any issues involved in multiprocessing since I'm kind of an expert at this point at multiprocessing/fork issues with Cygwin. I will note that the vast majority of problems can be resolved by properly rebasing all your DLLs.

DWesl commented:
The problem in `modfl` appears to affect only 64-bit Cygwin versions 3.1.5-3.1.7; versions 3.1.4 and earlier, as well as 3.2.0 and later, should both be fine.

I've only been testing this on 64-bit Cygwin, and the difficulty was primarily in trying to properly rebase the f2py-generated modules created as part of the testing process.  Not making an attempt to properly rebase these modules appears to be working for now.

DWesl committed:
Undef HAVE_MODFL only on 64-bit cygwin

32-bit modfl appears to work fine.  Problem exists in Cygwin 3.1.5 to 3.1.7 and should be fixed in 3.2.0.

charris commented:
What is the current status of this? Did #17548 have any effect?

DWesl commented:
I have similar code in `setup.py`, rather than `numpy/core/setup.py`, but I haven't actually tested with a compiler old enough to need the fix so I have no idea if it works as intended.  #17548 looks more precise, being used only for code that needs it, and on both Windows-native and cygwin GCC.  Should I drop my version?

charris commented:
> Should I drop my version and leave the other?

I don't know enough to make that decision.

DWesl commented:
For the broader question of how this is working; there were two tests that segfault on Cygwin versions between June and the present; we may want to wait until the fix is in a release before merging into main.

I apparently copied over the preview of Cygwin 3.2.0, so I still have segfaults in the `modfl` tests.  If anyone has suggestions for making a test not run so I can avoid `pytest-forked`, or a way to only fork some tests, that would be much appreciated.

DWesl commented:
> > Should I drop my version and leave the other?
>
> I don't know enough to make that decision.

The bulk of the changes in #17548 seems to be here:
https://github.com/numpy/numpy/pull/17548/files#diff-6fde01624b4d27874d419c0f8aeae3743c4f5e7e1c9f2b039eb453a714d0cbb1
which adds the `-ffixed-xmm%d` flags to compile lines with old gcc on Windows and Cygwin if targeting AVX512.

The corresponding code in this PR:
https://github.com/numpy/numpy/pull/16246/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R249
adds the `-ffixed-xmm` flags and `-fno-asynchronous-unwind-tables` to compile lines on 64-bit Cygwin.

Only one of `-ffixed-xmm` and `-fno-asynchronous-unwind-tables` are needed to fix the problem, and `-ffixed-xmm` is probably the right one to pick.  #17548 also seems to be more tightly targeted at the problem (processors without AVX512 don't run into this, but Windows probably would), and has definitely been tested for adding the flags to Numpy on its own.  I don't think the duplication is bad, but I'm not sure whether it's necessary.  I have no idea if there are things `setup.py` deals with that bypass `numpy/core/setup.py`, or the other way round.  I could also add comments to each point where flags are added pointing at the other place that adds flags, as well as the PRs in case the line numbers change.

charris commented:
@DWesl I'll be making a release in a day or two. You should pick what need to be added, if anything, and do it. There may, or may not, be another release in the 1.19 series. Have you tested with current master?

DWesl committed:
Revert "BLD: Fix flags added to compiler line for old GCC on MS."

This reverts commit 6ffbcb2e0df30863787aa3a3da796ad9a93c4a79.

Replaced by 17548.

DWesl committed:
Revert "BLD: Fix code checking for old GCC on cygwin."

This reverts commit d9c21bbffe176316192cdc4c7fd251ea99c7fecd.

Replaced by numpy/numpy#17548.

DWesl committed:
Revert "BLD: Remind GCC that .seh_savexmm fails for xmm16-31."

This reverts commit 92eaff35813f668a24cefd625600a80dbcd9e3c7.

Replaced by numpy/numpy#17548.

DWesl commented:
I dropped my attempt to add the compiler flags, since #17548 probably got tested and I don't think this did.

38 failures on master with python3.6, many due to the ufunc errors fixed/xfailed here, and more due to the polynomial class being confused about unicode vs ascii printing.

15 failures on master with python3.7, mostly due to the ufunc errors fixed or xfailed in this PR.  NumPy arrays are still marked as instances of Hashable, and there's a few other errors that aren't NumPy related.

7 failures on cygwin-fixes with python3.7, one due to not running `git clean -xdf`, one for a problem reported to Cygwin, and one that I can reproduce in python without NumPy, in addition to the `isinstance(np.arange(5), collections.Hashable)` failure mentioned earlier.

I can't think of anything else I know how to do, unless someone has pointers for fixing the `Hashable` thing.

charris committed:
Merge branch 'master' into cygwin-fixes

charris commented on numpy/core/include/numpy/npy_common.h:
```diff
@@ -10,6 +10,25 @@
+// int*, int64* should be propertly aligned on ARMv7 to avoid bus error
+#if !defined(NPY_STRONG_ALIGNMENT) && defined(__arm__) && !(defined(__aarch64__) || defined(_M_ARM64))
+#define NPY_STRONG_ALIGNMENT 1
+#endif
```
@seiko2plus @Qiyu8 Thoughts about this. Do you deal with this elsewhere?

seiko2plus replied:
yes, there's already an opened pr for it, see #18065

charris commented:
Closing, I have a rebased version that I will push for examination.

charris commented:
See #18102 for rebased version.
